tests/unit/apps/test_ovn.py

```diff
@@ -0,0 +1,208 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""Tests of the ovn application classes."""
+
+import pytest
+
+from cou.apps.ovn import OvnPrincipalApplication, OvnSubordinateApplication
+from cou.exceptions import ApplicationError
+from cou.steps import UpgradeStep
+from cou.utils import app_utils
+from cou.utils.openstack import OpenStackRelease
+from tests.unit.apps.utils import add_steps
+
+
+def test_ovn_principal(status, config, model):
+    app = OvnPrincipalApplication(
+        "ovn-central",
+        status["ovn_central_ussuri_22"],
+        config["auxiliary_ussuri"],
+        model,
+        "ovn-central",
+    )
+    assert app.channel == "22.03/stable"
+    assert app.os_origin == "distro"
+    assert app.apt_source_codename == "ussuri"
+    assert app.channel_codename == "yoga"
+    assert app.current_os_release == "yoga"
+
+
+def test_ovn_subordinate(status, model):
+    app = OvnSubordinateApplication(
+        "ovn-chassis",
+        status["ovn_chassis_ussuri_22"],
+        {},
+        model,
+        "ovn-chassis",
+    )
+    assert app.channel == "22.03/stable"
+    assert app.os_origin == ""
+    assert app.apt_source_codename is None
+    assert app.channel_codename == "yoga"
+    assert app.current_os_release == "yoga"
+
+
+def test_ovn_workload_ver_lower_than_22_principal(status, config, model):
+    target = "victoria"
+
+    exp_error_msg_ovn_upgrade = (
+        "OVN versions lower than 22.03 are not supported. It's necessary to upgrade "
+        "OVN to 22.03 before upgrading the cloud. Follow the instructions at: "
+        "https://docs.openstack.org/charm-guide/latest/project/procedures/"
+        "ovn-upgrade-2203.html"
+    )
+
+    app_ovn_central = OvnPrincipalApplication(
+        "ovn-central",
+        status["ovn_central_ussuri_20"],
+        config["auxiliary_ussuri"],
+        model,
+        "ovn-central",
+    )
+
+    with pytest.raises(ApplicationError, match=exp_error_msg_ovn_upgrade):
+        app_ovn_central.generate_upgrade_plan(target)
+
+
+def test_ovn_workload_ver_lower_than_22_subordinate(status, model):
+    target = "victoria"
+
+    exp_error_msg_ovn_upgrade = (
+        "OVN versions lower than 22.03 are not supported. It's necessary to upgrade "
+        "OVN to 22.03 before upgrading the cloud. Follow the instructions at: "
+        "https://docs.openstack.org/charm-guide/latest/project/procedures/"
+        "ovn-upgrade-2203.html"
+    )
+
+    app_ovn_chassis = OvnSubordinateApplication(
+        "ovn-chassis",
+        status["ovn_chassis_ussuri_20"],
+        {},
+        model,
+        "ovn-chassis",
+    )
+
+    with pytest.raises(ApplicationError, match=exp_error_msg_ovn_upgrade):
+        app_ovn_chassis.generate_upgrade_plan(target)
+
+
+def test_ovn_no_compatible_os_release(status, config, model):
+    ovn_central_status = status["ovn_central_ussuri_22"]
+    ovn_central_status.charm_channel = "55.7"
```

Comment:
nit: since we are extending the range to include lower versions, can we test that also? So parametrize the `charm_channel` value to one lower than the lower bound (e.g. 19.03) and one higher than the upper bound (e.g. 55.7).